### PR TITLE
attribute: add Set.String

### LINKED ISSUE #8209: attribute: add Set.String
Blocked by https://github.com/open-telemetry/opentelemetry-specification/pull/5110 (and stabilization)

Per https://github.com/open-telemetry/opentelemetry-go/pull/8205#issuecomment-4252990759

I think we should print it as a JSON object (map).

### PR DESCRIPTION
**Pull request by iyiola-dev**
Fixes #8209

Adds a [String](cci:1://file:///Users/iyiolaakanbi/projects/opentelemetry-go/attribute/kv.go:60:0-63:1) method to `attribute.Set` that returns the set as a JSON
object, consistent with the approach used by [Value.String](cci:1://file:///Users/iyiolaakanbi/projects/opentelemetry-go/attribute/value.go:323:0-362:1) and the in-progress
[KeyValue.String](cci:1://file:///Users/iyiolaakanbi/projects/opentelemetry-go/attribute/kv.go:60:0-63:1) (#8205).

For an empty or nil `*Set`, the method returns `{}`. Otherwise, it iterates
the set in its sorted order and writes `{"key":value,...}`, reusing the
existing [appendJSONString](cci:1://file:///Users/iyiolaakanbi/projects/opentelemetry-go/attribute/value.go:903:0-986:1) and [appendJSONValue](cci:1://file:///Users/iyiolaakanbi/projects/opentelemetry-go/attribute/value.go:854:0-901:1) helpers in [value.go](cci:7://file:///Users/iyiolaakanbi/projects/opentelemetry-go/attribute/value.go:0:0-0:0).

The returned string is meant for debugging; the string representation is
not stable.

**Benchmarks** (`go test -run=^$ -bench='^BenchmarkSetString$' -benchmem -count=3 ./attribute/`):

```console
goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/otel/attribute
cpu: Apple M1
BenchmarkSetString/Len2-8    17134850    77.36 ns/op     80 B/op    1 allocs/op
BenchmarkSetString/Len2-8    14539894    71.75 ns/op     80 B/op    1 allocs/op
BenchmarkSetString/Len2-8    17337637    69.29 ns/op     80 B/op    1 allocs/op
BenchmarkSetString/Len8-8     3627177    334.1 ns/op    320 B/op    1 allocs/op
BenchmarkSetString/Len8-8     3381171    348.5 ns/op    320 B/op    1 allocs/op
BenchmarkSetString/Len8-8     2950075    388.9 ns/op    320 B/op    1 allocs/op

### CODE REVIEW

**Review by pellared — DISMISSED** [review]
Blocked by the same reason as https://github.com/open-telemetry/opentelemetry-go/pull/8205 (by missing spec). 
See also: https://github.com/open-telemetry/opentelemetry-specification/pull/5028#discussion_r3258470518

> > An attribute set is analogous to a map. 
> 
> It is not specified anywhere. It a subject for a separate PR as we would need it for https://github.com/open-telemetry/opentelemetry-go/issues/8209. 

PS. Benchmarks are not readable can you fix it?

File: attribute/set.go [hunk 1]
```diff
@@ -406,6 +407,31 @@ func computeDataReflect(kvs []KeyValue) any {
...
+//
+// The returned string is meant for debugging;
+// the string representation is not stable.
+func (l *Set) String() string {
+	if l == nil || l.hash == 0 {
+		return "{}"
+	}
+
+	n := l.Len()
+	var b strings.Builder
```
Comment (pellared) [inline attribute/set.go:420]: This misses some `b.Grow` with at least some estimate.

Comment (iyiola-dev) [reply attribute/set.go:420]: yeah, this is done now @pellared

**Comment by iyiola-dev** [comment]
> Blocked by the same reason as #8205 (by missing spec). See also: [open-telemetry/opentelemetry-specification#5028 (comment)](https://github.com/open-telemetry/opentelemetry-specification/pull/5028#discussion_r3258470518)
> 
> > > An attribute set is analogous to a map.
> > 
> > 
> > It is not specified anywhere. It a subject for a separate PR as we would need it for #8209.
> 
> PS. Benchmarks are not readable can you fix it?

this is also fixed @pellared

**Comment by iyiola-dev** [comment]
hi @pellared can this be marked ready for review now or is this still blocked?

File: attribute/set.go [hunk 2]
```diff
@@ -406,6 +407,33 @@ func computeDataReflect(kvs []KeyValue) any {
...
+// String returns a string representation of the Set as a JSON object.
+//
+// The returned string is meant for debugging;
+// the string representation is not stable.
+func (l *Set) String() string {
+	if l == nil || l.hash == 0 {
+		return "{}"
+	}
+
+	n := l.Len()
```
Comment (dashpole) [inline attribute/set.go]: For an empty set created via `NewSet()`, `l.hash` is non-zero (`emptyHash`), so this currently allocates a `strings.Builder` to return `"{}"`. We can avoid the allocation by checking `if n == 0`:
```suggestion
	n := l.Len()
	if n == 0 {
		return "{}"
	}
```

File: attribute/set.go [hunk 3]
```diff
@@ -406,6 +407,33 @@ func computeDataReflect(kvs []KeyValue) any {
 	return at.Interface()
 }
 
+// String returns a string representation of the Set as a JSON object.
```
Comment (dashpole) [inline attribute/set.go]: Consider referencing the [OpenTelemetry Attribute Collection representation for non-OTLP](https://opentelemetry.io/docs/specs/otel/common/#attribute-collection-representation-for-non-otlp) specification rules here for consistency with `Value.String` and `KeyValue.String`.

File: attribute/set_test.go [hunk 4]
```diff
@@ -575,6 +575,94 @@ func generateStringAttrsWithSize(keyLen, valueLen int) []attribute.KeyValue {
...
+		{
+			name: "Nil",
+			set:  nil,
+			want: "{}",
+		},
+		{
+			name: "Empty",
+			set:  setPtr(attribute.NewSet()),
+			want: "{}",
+		},
```
Comment (dashpole) [inline attribute/set_test.go]: Could we also add test cases for the zero-value `attribute.Set{}` (to exercise the `l.hash == 0` branch) and an empty attribute value (`attribute.KeyValue{Key: "k"}` -> `{"k":null}`)?

**Comment by pellared** [comment]
> hi @pellared can this be marked ready for review now or is this still blocked?

Blocked by https://github.com/open-telemetry/opentelemetry-specification/pull/5149 and I am working on it.

In the meantime you can update the PR :wink:

**Comment by iyiola-dev** [comment]
hi @pellared i have updated this and i have marked it ready for review.

File: attribute/benchmark_test.go [hunk 5]
```diff
@@ -585,3 +585,55 @@ func benchmarkEquivalentMapAccess(b *testing.B, set *attribute.Set) {
...
+			attribute.String("b", "two"),
+			attribute.String("c", "three"),
+			attribute.String("d", "four"),
+			attribute.String("e", "five"),
+			attribute.String("f", "six"),
+			attribute.String("g", "seven"),
+			attribute.String("h", "eight"),
+			attribute.String("i", "nine"),
+			attribute.String("j", "ten"),
+		)
```
Comment (pellared) [inline attribute/benchmark_test.go]: Let's just do "Len2" and "Len8" like in other benchmarks.

File: attribute/benchmark_test.go [hunk 6]
```diff
@@ -585,3 +585,55 @@ func benchmarkEquivalentMapAccess(b *testing.B, set *attribute.Set) {
...
 }
+
+func BenchmarkSetString(b *testing.B) {
+	b.Run("Empty", func(b *testing.B) {
+		set := attribute.NewSet()
+		b.ReportAllocs()
+		b.ResetTimer()
+		for range b.N {
+			outStr = set.String()
+		}
```
Comment (pellared) [inline attribute/benchmark_test.go]: This can be


```suggestion
		for b.Loop() {
			set.String()
		}
```

Please apply this to other cases in this PR.

File: attribute/set_test.go [hunk 7]
```diff
@@ -595,6 +596,117 @@ func generateStringAttrsWithSize(keyLen, valueLen int) []attribute.KeyValue {
...
+		want string
+	}{
+		{
+			name: "Nil",
+			set:  nil,
+			want: "{}",
+		},
+		{
+			name: "ZeroValue",
+			set:  new(attribute.Set{}),
```
[the remark this answers is not in the document]
Comment (pellared) [reply attribute/set_test.go:612]: The comment is a false positive. It complies since Go 1.26. See https://go.dev/doc/go1.26#language

**Comment by iyiola-dev** [comment]
comments treated @pellared

**Comment by pellared** [comment]
> comments treated @pellared

Can you update the PR description with latest benchmark results?

**Review by pellared — APPROVED** [review]

**Comment by iyiola-dev** [comment]
> > comments treated @pellared
> 
> Can you update the PR description with latest benchmark results?

done @pellared

**Review by dmathieu — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]
